### PR TITLE
Add `dhall-to-nixpkgs`

### PR DESCRIPTION
**Pull request by Gabriel439**
This adds a new `dhall-nixpkgs` project which builds a `dhall-nixpkgs`
executable.  This executable is the Dhall analog of `cabal2nix`, which
takes a directory or GitHub project as input and generates a buildable
Nix package as its output.

Example output:

```bash
$ dhall-to-nixpkgs github https://github.com/Gabriel439/dhall-semver.git
```
```nix
{ buildDhallGitHubPackage, Prelude }:
  buildDhallGitHubPackage {
    name = "dhall-semver";
    githubBase = "github.com";
    owner = "Gabriel439";
    repo = "dhall-semver";
    rev = "7f5b7bf0e89664571522de9919edcc8469a55320";
    fetchSubmodules = false;
    sha256 = "00vi74npwbhwn2lnjpw4zv5y9kjqysc2f1wg1zncps6k5wljnwb7";
    directory = "";
    file = "package.dhall";
    source = false;
    dependencies = [ (Prelude.override { file = "package.dhall"; }) ];
    }
```

```bash
$ dhall-to-nixpkgs directory ~/proj/dhall-packages/
```
```nix
{ buildDhallDirectoryPackage, dhall-kubernetes, Prelude, dhall-packages }:
  buildDhallDirectoryPackage {
    name = "dhall-packages";
    src = /home/gabriel/proj/dhall-packages;
    file = "package.dhall";
    source = false;
    dependencies = [
      (dhall-kubernetes.override { file = "1.15/package.dhall"; })
      (Prelude.override { file = "JSON/Type"; })
      (Prelude.override { file = "JSON/package.dhall"; })
      (dhall-kubernetes.override { file = "1.16/package.dhall"; })
      (dhall-kubernetes.override { file = "1.17/package.dhall"; })
      (dhall-kubernetes.override {
        file = "types/io.k8s.api.core.v1.ServiceSpec.dhall";
        })
      (dhall-kubernetes.override {
        file = "types/io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta.dhall";
        })
      (dhall-kubernetes.override {
        file = "types/io.k8s.apimachinery.pkg.apis.meta.v1.Time.dhall";
        })
      (dhall-kubernetes.override {
        file = "types/io.k8s.api.core.v1.PodTemplateSpec.dhall";
        })
      (dhall-kubernetes.override {
        file = "types/io.k8s.api.core.v1.SecretKeySelector.dhall";
        })
      (dhall-kubernetes.override {
        file = "types/io.k8s.apimachinery.pkg.apis.meta.v1.MicroTime.dhall";
        })
      (dhall-kubernetes.override {
        file = "types/io.k8s.apimachinery.pkg.apis.meta.v1.ListMeta.dhall";
        })
      (Prelude.override { file = "Map/Type"; })
      (dhall-packages.override { file = "kubernetes/k8s/1.15.dhall"; })
      (Prelude.override { file = "package.dhall"; })
      ];
    }
```

This requires the following matching changes to Nixpkgs in order to work:

* Two new `buildDhall{Directory,GitHub}Package` utilities have to be
  added

  ... which are used by the generated output

* The Dhall package set in Nixpkgs now needs to select default versions
  for packages

  In other words, `pkgs.dhallPackages.Prelude` should select a default
  version of the Prelude (ideally the latest version)

  This parallels how the Haskell package set works, by selecting a
  default version for each package, the latest version possible that
  produces a consistent package set.

### CODE REVIEW

**Comment by Gabriel439** [comment]
Here's the matching change to Nixpkgs: https://github.com/NixOS/nixpkgs/pull/89237

File: dhall-nixpkgs/Main.hs [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+    <https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/dhall/build-dhall-package.nix>
+
+    The Nixpkgs support for Dhall essentially replaces all remote imports with
+    cache hits, but doing so implies that all remote imports must be protected
+    by an integrity check.
+
+    This function finds all remote imports that are transitive dependencies of
+    the given expression, failing if any of them are missing integrity checks.
+-}
+findExternalDependencies :: Expr Src Import -> StateT Status Shell URL
```
Comment (sjakobi) [inline dhall-nixpkgs/Main.hs:256]: What's the purpose of the `Shell` monad here? I don't quite understand how it's being used here.

File: dhall-nixpkgs/Main.hs [hunk 2]
```diff
@@ -0,0 +1,828 @@
...
+    names and files that follow this convention.  For example, given a Dhall
+    import like:
+
+        https://raw.githubusercontent.com/EarnestResearch/dhall-packages/master/kubernetes/k8s/1.14.dhall
+
+    ... this will create the corresponding Nix dependency of the form:
+
+        dhall-packages.override { file = "kubernetes/k8s/1.14.dhall"; }
+-}
+dependencyToNix :: URL -> IO ((Text, Maybe nExpr), NExpr)
```
Comment (sjakobi) [inline dhall-nixpkgs/Main.hs:344]: Could you document the components of the result type?

File: dhall-nixpkgs/Main.hs [hunk 3]
```diff
@@ -0,0 +1,828 @@
...
+
+    let prelude = "Prelude"
+
+    case authority of
+        "raw.githubusercontent.com" -> do
+            let File{..} = path
+
+            let Dhall.Core.Directory{..} = directory
+
+            case reverse (file : components) of
```
Comment (sjakobi) [inline dhall-nixpkgs/Main.hs:356]: Aside/style: I _think_ I understand this because I have seen the `File` and `Directory` types multiple times. I believe that other readers will have more difficulty telling where `directory`,  `file` and `components` come from.

I've never quite understood why you favor this style. :)

IMHO this could be made clearer in multiple ways: By using puns instead of wildcards, by simply using the record accessors, or with `lens`'s `view`, …

File: dhall-nixpkgs/Main.hs [hunk 4]
```diff
@@ -0,0 +1,828 @@
...
+                -- Special case to recognize a Prelude import and treat it as if
+                -- it were an import of prelude.dhall-lang.org
+                "dhall-lang" : "dhall-lang" : _rev : "Prelude" : rest -> do
```
Comment (sjakobi) [inline dhall-nixpkgs/Main.hs:359]: `_rev` is "forgotten", right? What's the rationale for that?

File: dhall-nixpkgs/Main.hs [hunk 5]
```diff
@@ -0,0 +1,828 @@
...
+
+Dhall's Nixpkgs requires that all of your remote dependencies are protected by
+by semantic integrity checks.  This ensures that Nix can replace the remote
+imports with cached imports built by Nix instead of the imports being fetched
+via HTTP requests using Dhall.
+
+The following dependency is missing a semantic integrity check:
+
+↳ $dependency
+|]
```
Comment (sjakobi) [inline dhall-nixpkgs/Main.hs:315]: IMHO the lack of indentation makes it a bit tricky to understand the structure of the following code. Maybe move these blobs to the top level?

File: dhall-nixpkgs/Main.hs [hunk 6]
```diff
@@ -0,0 +1,828 @@
+{-# LANGUAGE ApplicativeDo         #-}
```
[the remark this answers is not in the document]
Comment (Gabriel439) [reply dhall-nixpkgs/Main.hs:1]: I haven't had issues with `ApplicativeDo`.  I use it all the time at work

File: dhall-nixpkgs/Main.hs (hunk 1 above)
Comment (Gabriel439) [reply dhall-nixpkgs/Main.hs:302]: @sjakobi: The `Turtle.Shell` module explains a bit how it works:

http://hackage.haskell.org/package/turtle-1.5.19/docs/Turtle-Shell.html

... as well as the `turtle` tutorial:

http://hackage.haskell.org/package/turtle-1.5.19/docs/Turtle-Tutorial.html#g:9

File: dhall-nixpkgs/Main.hs (hunk 4 above)
Comment (Gabriel439) [reply dhall-nixpkgs/Main.hs:402]: I just added a long module header which explains at the end why `dhall-to-nixpkgs` discards the package version / revision when generating the dependency list

**Comment by Gabriel439** [comment]
At this point I'm mainly waiting on https://github.com/NixOS/nixpkgs/pull/89237 to be approved, too, before merging this

File: dhall-nixpkgs/Main.hs [hunk 7]
```diff
@@ -0,0 +1,948 @@
...
+                @@  Nix.attrsE
+                        [ ("name", Nix.mkStr finalName)
+                        , ("src", Nix.mkPath False src)
+                        , ("file", Nix.mkStr (Turtle.format fp file))
+                        , ("source", Nix.mkBool source)
+                        , ("dependencies", Nix.mkList (nub (fmap dependencyExpression nixDependencies)))
+                        ]
+                )
+
+    Prettyprint.Text.putDoc (Nix.Pretty.prettyNix nixExpression)
```
Comment (sjakobi) [inline dhall-nixpkgs/Main.hs:676]: Does `prettyNix` append a newline? `putDoc` doesn't seem to…

File: dhall-nixpkgs/Main.hs [hunk 8]
```diff
@@ -0,0 +1,948 @@
...
+
+Dhall's Nixpkgs requires that all of your remote dependencies are protected by
+by semantic integrity checks.  This ensures that Nix can replace the remote
+imports with cached imports built by Nix instead of the imports being fetched
+via HTTP requests using Dhall.
+
+The following dependency is missing a semantic integrity check:
+
+↳ $dependency
+|]
```
Comment (sjakobi) [inline dhall-nixpkgs/Main.hs:717]: You're not indenting the text to avoid overlong lines, right?

Comment (Gabriel439) [reply dhall-nixpkgs/Main.hs:717]: @sjakobi: It's mostly because it makes it easier to ensure that the message fits in 80 columns

**Comment by Anton-Latukha** [comment]
You can review this code later. HNix is 0.8.0, and it now uses GHC 8.8 by default: https://matrix.hackage.haskell.org/#/package/hnix

**Comment by PierreR** [comment]
@Gabriel439 did something change on the command line prompt ?
I am trying the suggested:
```
dhall-to-nixpkgs github https://github.com/Gabriel439/dhall-semver.git
dhall-to-nixpkgs: nix-prefetch-git: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
```
